File: DFS/79.word-search.go
package leetcode

// @leet start
type Pt struct {
	r, c int
}

// move returns a new point by moving p by the offset d.
func (p Pt) move(d Pt) Pt {
	return Pt{r: p.r + d.r, c: p.c + d.c}
}

func exist(board [][]byte, word string) bool {
	nr := len(board)
	nc := len(board[0])

	dirs := []Pt{
		{0, 1},
		{0, -1},
		{-1, 0},
		{1, 0},
	}
	visited := make(map[Pt]bool)

	// dfs is a recursive helper function that searches for the word.
	var dfs func(p Pt, step int) bool
	dfs = func(p Pt, step int) bool {
		if p.r < 0 || p.r >= nr || p.c < 0 || p.c >= nc {
			return false
		}
		if board[p.r][p.c] != word[step] || visited[p] {
			return false
		}
		if step == len(word)-1 {
			return true
		}

		visited[p] = true

		for _, d := range dirs {
			if dfs(p.move(d), step+1) {
				return true
			}
		}

		visited[p] = false
		return false
	}

	// Try starting DFS from every cell that matches the first letter.
	for r := 0; r < nr; r++ {
		for c := 0; c < nc; c++ {
			if board[r][c] == word[0] {
				if dfs(Pt{r, c}, 0) {
					return true
				}
			}
		}
	}
	return false
}

// @leet end

File: DFS/79.word-search.py
from typing import List, Self


# @leet start
from dataclasses import dataclass


@dataclass(frozen=True)
class pt:
    r: int
    c: int

    def move(self, d: Self):
        return pt(r=self.r + d.r, c=self.c + d.c)


class Solution:
    dirs = [
        pt(0, 1),
        pt(0, -1),
        pt(-1, 0),
        pt(1, 0),
    ]

    def exist(self, board: List[List[str]], word: str) -> bool:
        nr, nc = len(board), len(board[0])

        def get(p: pt):
            return board[p.r][p.c]

        def outSide(p: pt):
            return not (p.r >= 0 and p.r < nr and p.c >= 0 and p.c < nc)

        seen = set()

        def dfs(p: pt, step: int) -> bool:
            if outSide(p) or pt in seen or get(p) != word[step]:
                return False
            if step == len(word) - 1:
                return True
            seen.add(pt)
            for d in self.dirs:
                if dfs(p.move(d), step + 1):
                    return True
            seen.remove(pt)
            return False

        for r in range(nr):
            for c in range(nc):
                p = pt(r, c)
                if dfs(p, 0):
                    return True
        return False


# @leet end

